### ISSUE TITLE
sql: ON DELETE SET DEFAULT violates constraints

### ISSUE THREAD

**Issue by knz** [cockroachdb/cockroach#23361]
Found while investigating #22304.

```sql
 create table base (x int primary key);
 create table derived(x int primary key, y int default 123 references base(x) on delete set default);
 insert into base values(456);
 insert into derived values (111, 456);

 --- bang:
 delete from base where true;
```

Expected, pg has this correct:
```
pq: foreign key violation: value [123] not found in base@primary [x]
```

in CockroachDB, currently, incorrect: no error.

cc @BramGruneir.

I have a patch incoming.

**Comment by knz** [comment]
(Found because some logic tests in the `cascade` file are exploiting the incorrect behavior and they broke after I inadvertently fixed this bug)

**Comment by knz** [comment]
Note: the error is in the DELETE code, not the FK code.

**Comment by BramGruneir** [comment]
I thought I checked for that case.  Doesn't `subtest DeleteSetDefault_Basic1_WrongDefault` cover this case?

**Comment by knz** [comment]
Check out https://github.com/cockroachdb/cockroach/blob/master/pkg/sql/logictest/testdata/logic_test/cascade#L2734-L2771

After the TRUNCATE, table `a` is empty and does not contain the default values for the `b` tables any more. The ON DELETE SET DEFAULT should fail with an error.

**Comment by BramGruneir** [comment]
Oh wow!  I can't believe I missed that!  Good catch!

**Comment by knz** [comment]
This has been fixed by #23657 or some other patch related to #23373 already.
